runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java

```diff
@@ -236,14 +236,26 @@ public long getConflictSetAndCommit(ConflictSetInfo conflictSet) {
         }
 
         // If the write set is empty, this is a read-only transaction.
-        // Return the max address of all accessed streams, this will provide a safe token of the
+        // If the transaction has only read non-monotonic objects, then return the
+        // max address of all accessed streams, this will provide a safe token of the
         // transaction's snapshot. Notice that, providing the snapshot of the tx (vs. max address)
         // can lead to data loss, as a sequencer reboot might incur in sequence regression.
         // This timestamp is aimed to provide clients with a secure point for delta/streaming
         // subscription, the later could lead to data loss scenarios.
+        // If the transaction has read monotonic objects, we instead return the min address
+        // of all accessed streams. Although this avoids data loss, clients subscribing at
+        // this point for delta/streaming may observe duplicate data.
         if (getWriteSetInfo().getWriteSet().getEntryMap().isEmpty()) {
             log.trace("Commit[{}] Read-only commit (no write)", this);
-            return getMaxAddressRead();
+
+            final boolean hasReadMonotonicObject = knownStreamsPosition.keySet()
+                    .stream().anyMatch(proxy -> proxy.getUnderlyingObject().isMonotonicObject());
+
+            if (hasReadMonotonicObject) {
+                return getMinAddressRead();
```

Comment:
What if you are operating only on monotonic objects or a single monotonic object?

* TX1 starts the full sync by invoking `entryStream()`.
* TX2 comes along and does a read. There were new updates in the meanwhile, so the known tails is going to be advanced.
* TX1 finishes the full sync and does a `commit()`, which will return the address of TX2 stream position.

Is seems to me that there is still a gap.

Reply:
`knownStreamsPosition` is specific to each transaction. Hence TX1 will not update the stream position. It is only updated if TX1 reads the stream again.